### PR TITLE
Logging configuration through ASPIRE config file (`confuse`)

### LINKED ISSUE #642: User Logging Config
User requested we expose and/or document configuring logging.

It has been a while since we reviewed our logging system, but I was probably the last one to touch it...

As I recall, currently we are just using the standard Python logging module, configured using `logging.conf` in our repo root.  The log stream is split and filtered separately for the console (INFO) and disk (DEBUG). 

It is easy to modify that as a dev, but for users here are a few ideas:

- Tutorial showing how to change level in a script or notebook
- Flags in our command line utils to set it.
- Replace some noisy logic (batching for example) with tqdm style progress bars to convey the same information.
- Perhaps revisit the logging config as it relates to #504 
- Could ask the RSE group what they're doing re best practice for user configurable Python logging.

Some of these are things we can do now, while others might be medium sized efforts.

### PR DESCRIPTION
**Pull request by chris-langfield**
See #642 

This PR adds 3 new options to the `confuse` config: ~~`logging.log_file`~~, `logging.log_prefix`, `logging.console_level`, and `logging.log_file_level`. 

These are passed through to `logging.conf` via interpolation. This means that that values from our config will always supersede Python defaults. However, the `logging.conf` stays so that advanced users can mess with the formatter, and other more sophisticated options. 

Still to do: 
- [x] logging tutorial
- [x] command line option

### CODE REVIEW

**Comment by chris-langfield** [comment]
Config page with logging tutorial:

[ASPIRE Configuration — ASPIRE 0.10.1 documentation.pdf](https://github.com/ComputationalCryoEM/ASPIRE-Python/files/10136537/ASPIRE.Configuration.ASPIRE.0.10.1.documentation.pdf)

**Comment by chris-langfield** [comment]
I intended to change the log file timestamp to be `YYYY-MM-DDThh:mm:ss` (colons instead of dashes) to be more human readable but it looks like this breaks the Windows tests hah. Gonna look into it

**Comment by chris-langfield** [comment]
Yeah looks like it's possible to replace it with a unicode character that looks like a colon, but I'm just gonna turn it back into a dash

File: src/aspire/__init__.py [hunk 1]
```diff
@@ -23,14 +23,24 @@
 # We'll reassign the evaluated log_dir back into the config so it displays well.
 config["logging"]["log_dir"] = log_dir_path.as_posix()
 
+# log output file prefix
+log_prefix = config["logging"]["log_prefix"].get()
```
Comment (garrettwrong) [inline src/aspire/__init__.py]: enforce a type here and the `.get()`s below.

File: src/aspire/__init__.py [hunk 2]
```diff
@@ -23,14 +23,24 @@
...
+
+# DEBUG, INFO, etc.
+console_level = config["logging"]["console_level"].get()
```
Comment (garrettwrong) [inline src/aspire/__init__.py]: Maybe we can use `as_choice(choices)` here to enforce the value is sane.

File: src/aspire/commands/__init__.py [hunk 3]
```diff
@@ -0,0 +1,13 @@
...
+
+from aspire import config
+
+# universal option for aspire commands
+log_level_option = click.option(
+    "--loglevel",
+    type=click.Choice(
+        ["DEBUG", "INFO", "WARNING", "ERROR", "CRITICAL"], case_sensitive=False
+    ),
+    default=config["logging"]["console_level"].get(),
```
Comment (garrettwrong) [inline src/aspire/commands/__init__.py]: get type

File: src/aspire/commands/apple.py [hunk 4]
```diff
@@ -81,9 +83,12 @@ def apple(
     response_thresh_norm_factor,
     conv_map_nthreads,
     n_processes,
+    loglevel,
 ):
     """Pick and save particles from one or more mrc files."""
 
+    logger.setLevel(getattr(logging, loglevel))
```
Comment (garrettwrong) [inline src/aspire/commands/apple.py]: I think this line and all the others like it does not do what you think.

**Review by garrettwrong — CHANGES_REQUESTED** [review]
Logging changes look good, just some minor type enforcement there.

I didn't check it out to run it, but unless I am misreading I don't believe the CLI changes here are correct.  Double check that and we can discuss if needed.

File: src/aspire/commands/__init__.py (hunk 3 above)
Comment (chris-langfield) [reply src/aspire/commands/__init__.py]: Done

File: src/aspire/__init__.py (hunk 2 above)
Comment (chris-langfield) [reply src/aspire/__init__.py]: Done

File: src/aspire/__init__.py (hunk 1 above)
Comment (chris-langfield) [reply src/aspire/__init__.py]: Done

File: src/aspire/commands/apple.py (hunk 4 above)
Comment (chris-langfield) [reply src/aspire/commands/apple.py]: You're totally right. I realized the important thing is setting the level of the `StreamHandler` of the *root* logger. Added utility functions for this. Tested it a few times and works (in particular, all messages still sent to log file)

**Comment by chris-langfield** [comment]
- Added utility functions which get and set **root** logger handler levels to control severity of messages that get out into console or file.
- `tqdm` and `trange` should not display unless console level is `INFO` or lower
- Added the name of the module that originated the message to the logger's formatter.

Some notes on root vs module level loggers:
- access root logger via `logging.getLogger()`
- **module**-level loggers accessed via `logging.getLogger(__name__)` in a module can still be individually adjusted. For example, if I set `logger.setLevel("WARNING")` within `aspire.source.image` and do nothing else, `INFO` messages will still print to console as normal, unless they originated from `aspire/source/image.py`.
- This will filter which messages are passed up the handlers (which belong to the root logger)

**Review by garrettwrong — DISMISSED** [review]
Thanks for making those changes!

File: src/aspire/utils/logging.py [hunk 5]
```diff
@@ -81,5 +85,63 @@ def trange(*args, **kwargs):
...
+        getConsoleLoggingLevel() not in ["DEBUG", "INFO"]
+    )
     return _tqdm.trange(*args, **kwargs, disable=disable)
+
+
+def setConsoleLoggingLevel(level_name):
+    """
+    Dynamically sets the console logging level by setting the level of the root logger's StreamHandler to `level_name`.
+    Note this will supersede the `logging.console_level` option stored in ASPIRE's configuration file.
+    :param level_name: One of "DEBUG", "INFO", "WARNING", "ERROR", "CRITICAL".
```
Comment (j-c-c) [inline src/aspire/utils/logging.py]: Need a blank line between description and `param`.

File: src/aspire/utils/logging.py [hunk 6]
```diff
@@ -81,5 +85,63 @@ def trange(*args, **kwargs):
...
+    stream_handler = logging.getLogger().handlers[0]
+    stream_handler.setLevel(getattr(logging, level_name))
+
+
+def getConsoleLoggingLevel():
+    """
+    Returns the Python logging level of the root logger's StreamHandler, i.e. console output.
+    This is the same as the `logging.console_level` option in ASPIRE's configuration file unless
+    the console logging level has been changed dynamically during a session. (e.g. by a CLI option)
+    :return: The current console logging level name as a string. One of "DEBUG". "INFO", "WARNING",
```
Comment (j-c-c) [inline src/aspire/utils/logging.py]: blank line between description and `return`

File: src/aspire/utils/logging.py [hunk 7]
```diff
@@ -81,5 +85,63 @@ def trange(*args, **kwargs):
...
+    # handler list is ordered according to logging.conf
+    stream_handler = logging.getLogger().handlers[0]
+    return logging.getLevelName(stream_handler.level)
+
+
+def setFileLoggingLevel(level_name):
+    """
+    Dynamically sets the log file logging level by setting the level of the root logger's FileHandler to `level_name`.
+    Note this will supersede the `logging.log_file_level` option stored in ASPIRE's configuration file.
+    :param level_name: One of "DEBUG", "INFO", "WARNING", "ERROR", "CRITICAL".
```
Comment (j-c-c) [inline src/aspire/utils/logging.py]: Same, add blank line.

File: src/aspire/utils/logging.py [hunk 8]
```diff
@@ -81,5 +85,63 @@ def trange(*args, **kwargs):
...
+    file_handler.setLevel(getattr(logging, level_name))
+
+
+def getFileLoggingLevel():
+    """
+    Returns the Python logging level of the root logger's FileHandler, i.e. log file output.
+    This is the same as the `logging.log_file_level` option in ASPIRE's configuration file
+    unless the file logging level has been changed dynamically during a session. (e.g. by a CLI option)
+    :return: The current file logging level name as a string. One of "DEBUG", "INFO", "WARNING",
+    "ERROR", "CRITICAL".
```
Comment (j-c-c) [inline src/aspire/utils/logging.py]: Blank line above `return` and indent the line wrap.

File: gallery/tutorials/configuration.py [hunk 9]
```diff
@@ -120,6 +120,60 @@
...
+# as in the example above):
+#
+#    .. code-block:: yaml
+#
+#      logging:
+#          log_dir=/tmp/my_proj/aspire_logs
+#
+# ASPIRE directs output streams to both the console and to a log file during each session. The logging verbosity for both of these
+# can be specified individually via ``console_level`` and ``log_file_level``. These levels are passed through to Python's logging
+# system, so the labels are the same: ``DEBUG, INFO, WARNING, ERROR, CRITICAL``. In general, it may make sense to have the log file
```
Comment (j-c-c) [inline gallery/tutorials/configuration.py]: Maybe separate quotes for each label: ``DEBUG``, ``INFO``, ``WARNING``, ``ERROR``, ``CRITICAL``

**Review by j-c-c — COMMENTED** [review]
Looks good! Just a few things.

File: src/aspire/utils/logging.py (hunk 6 above)
Comment (chris-langfield) [reply src/aspire/utils/logging.py]: thanks for catching, done

**Comment by chris-langfield** [comment]
Thanks for catching those docstrings

File: gallery/tutorials/configuration.py (hunk 9 above)
Comment (chris-langfield) [reply gallery/tutorials/configuration.py]: Sure, done

**Review by j-c-c — APPROVED** [review]
Looks good!

**Comment by garrettwrong** [comment]
I think since we discussed this at our meeting with positive feedback that you may merge it when you are ready.  Thanks!

**Review by garrettwrong — APPROVED** [review]

**Pull request merged**